---
title: "Heatmap Generator Database Guide"
author: "Lucas Gray"
date: "Friday, May 18, 2016"
output: html_document
runtime: dplyr
---
### <a name="top"></a> Introduction and Requirements

This guide will describe the database tables used by our shiny Heatmap Generator application, and will help you build and update these databases so that you have the most up-to-date data and results presented interactively.

The Database structure is built around 3 tables:

* [*data*](#rpkm_section) - contains the data that will be presented in the heatmap
* [*desc*](#desc_section) - describes the annotations that will be displayed for each cell
* [*anno*](#anno_section) - contains the annotations for each cell:
    + [Categorical](#categorical) data like clusters and Cre lines
    + [Quantitative](#quantitative) data like total reads mapped

Once you've built these 3 tables, you can use scrattch helper functions to write or update the database, and check the structure:

* [*Writing the Database*](#db_write_section) - initializing a database with write_database()
* [*Updating a Database*](#db_update_section) - updating individual tables with write_database()
* [*Check Database Structure*](#db_write_section) - checking the structure with check_db_structure()

To build or modify the database, you will need the DBI and RSQLite packages installed. These should be added as dependencies for scrattch.

I make heavy use of the dplyr package in my scripting, so if the code below looks unfamiliar, take a look at dplyr. dplyr adds many convenient functions, and makes use of the magrittr package to allow functions to be strung together. A tutorial for dplyr can be found [here](https://cran.rstudio.com/web/packages/dplyr/vignettes/introduction.html).

I also prefer that stringsAsFactors=F be specified as an option at the start of any script.

```{r, eval=FALSE}
library(scrattch)
library(dplyr)

options(stringsAsFactors=F)
```

```{r, include=FALSE}
library(scrattch)
library(dplyr)

options(stringsAsFactors=F)
```

-----

### <a name="rpkm_section"></a> The *data* table

The data table has the same structure as the rpkm.csv table that is an output of our RNA-Seq analysis pipeline: Each row is a gene, and each column is a cell. For demonstration purposes, here is a very small sample of our cortical V1 RPKM data:

```{r}
rpkm <- v1_data[1:5,1:11]
rpkm[1:5]
```

To prepare the data for the database, note that the data table should be a data frame with the cell IDs used as column names, and the gene names should be in the first column, named "genes", and have the class "character":
```{r}
names(rpkm)
str(rpkm[1:5])
```

[Back to Top](#top)

-----

### <a name="desc_section"></a> The *desc* Table

The *desc* table describes the annotations that are in the annotations (anno) table. This table contains the base of the annotation column names and the label that will be used for the annotation in some functions.

For example, if the annotations are WGCNA clusters, Cre lines, and Number of Mapped Reads, this table could be easily made in R:

```{r}
desc <- data.frame(base=c("wgcna","cre","mapped"),
                   name=c("WGCNA Clusters","Cre Line","Number of Mapped Reads"))

desc
```

For demonstration purposes, we've also supplied a pre-built desc table for the mouse V1 RPKM data

```{r}
v1_desc
```


[Back to Top](#top)

-----

### <a name="anno_section"></a> The *anno* Table

The *anno* table contains the annotations that are used to group and arrange data from each cell in the heatmap generator.

In the *anno* table, each row represents a cell. The first column of the table contains cell IDs that match the column names of the *rpkm* table.

Subsequent columns represent each annotation with three columns:

* _id: A numeric column that determines the order of the groups in the annotation
* _label: A character column that has the name of the annotation group, or the value of the annotation, in the case of quantitative values
* _color: A character column with the color of the annotation group as either a [hex color](http://www.color-hex.com/) or an [R color](http://www.stat.columbia.edu/~tzheng/files/Rcolor.pdf)

This table can be built in Excel as a csv, then imported into the database. This table is a representative set for the 10 cells included in the *rpkm* table demo. If your table already has this format, it can be written directly to the database:

```{r}
complete.anno <- v1_anno[1:10,]
head(complete.anno)
```

This table can also be built in R, or tables with partial data can be supplemented in R to include all three required columns. Below, a table with WGCNA clusters, Cre lines, and Mapped read values are expanded to include \_id and \_color columns.

```{r}
partial.anno <- demo_anno
names(partial.anno) <- c("sample_id","wgcna_label","cre_label","mapped_label")

head(partial.anno)
```

<a name="categorical"></a>To fill out the \_id and \_color columns, a summary table can be made for each annotation. These can be arranged as desired, colors can be added, and then the data can be joined together to make a complete *anno* table.

Below, I use the WGCNA Cluster labels to generate wgcna_id and wgcna_color data.
```{r}
wgcna.anno <- partial.anno %>% select(wgcna_label)
head(wgcna.anno)

wgcna.anno <- wgcna.anno %>%
  unique() %>%
  arrange(wgcna_label) %>%
  mutate(wgcna_id=1:n(),
         wgcna_color=rainbow(n()))

wgcna.anno

```

Colors and ids could also be specified explicitly to define the order and color:
```{r}
cre.anno <- data.frame(cre_id=1:3,
                       cre_label=c("Gad2","Rorb","Ntsr1"),
                       cre_color=c("orangered","dodgerblue","mediumorchid4"))

cre.anno
```

<a name="quantitative"></a>For quantitative data, ids should be assigned based on the order in which you would like the data to be arranged, and colors can be assigned by choosing a position within a color vector. I use colorRampPalette() to define a vector of 1000 colors along a gradient, then use some simple math to decide which color to pull from the vector based on the maximum value for the annotation:
```{r}
color.space <- colorRampPalette(c("darkblue","dodgerblue","gray20","orangered","red"))(1000)

mapped.anno <- partial.anno %>% 
  select(mapped_label) %>% 
  arrange(mapped_label) %>%
  mutate(mapped_id=1:n(),
         mapped_color=color.space[round(mapped_label/max(mapped_label)*1000,0)])

mapped.anno
```

Now, all of these values can be joined together with the original, partial annotations to build the final anno table:
```{r}
anno <- partial.anno %>%
  left_join(wgcna.anno) %>%
  left_join(cre.anno) %>%
  left_join(mapped.anno)

head(anno)
```

This table is now complete and can be added to the database. It's not required, but you may want to arrange the table to group the annotation columns together:
```{r}
anno <- anno %>%
  select(sample_id,
         wgcna_id,wgcna_label,wgcna_color,
         cre_id,cre_label,cre_color,
         mapped_id,mapped_label,mapped_color) %>%
  mutate(mapped_label = as.character(mapped_label))

head(anno)
```

The database now contains all three required tables, and should work with the scrattch plotting functions.

[Back to Top](#top)

