---
title: "scrattch Feather Format Guide"
author: "Lucas Gray"
date: "Tuesday, Jun 6, 2017"
output: html_document
runtime: dplyr
---
### <a name="top"></a> Introduction and Requirements

This guide will describe the feather tables used by scrattch for interactive data visualization in our Shiny apps, and will help you build and update these files so that you have the most up-to-date data and results presented interactively.

### < a name="structure"></a> Feather file structure

The feather file structure is built around 3 required tables:

* [*data*](#data_section) - contains the data that will be presented in the heatmap
* [*desc*](#desc_section) - describes the annotations that will be displayed for each cell
* [*anno*](#anno_section) - contains the annotations for each cell:
    + [Categorical](#categorical) data like clusters and Cre lines
    + [Quantitative](#quantitative) data like total reads mapped or genes detected

These optional tables/RData objects enhance visualization and analysis:

* [*data_t*](#data_t_section) - A transposed data table for fast retrieval of all gene data per sample
* [*dend*](#dend_section) - an RData file that contains a dendrogram object
* [*memb*](#memb_section) - Node membership scores used for cell that are mapped onto a taxonomy (e.g. Patch-seq results)
* [*tsne*](#tsne_section) - tSNE dimensions for apps that plot tSNE data
* [*tsne_desc*](#tsne_desc_section) - Descriptions of the tSNE dimension sets
* [*prop*](#prop_section) - Proportions of cells in each cluster that express each gene, used by the Sifter app to find marker genes

I make heavy use of the dplyr package in my scripting, so if the code below looks unfamiliar, take a look at dplyr. dplyr adds many convenient functions, and makes use of the magrittr package to allow functions to be strung together. A tutorial for dplyr can be found [here](https://cran.rstudio.com/web/packages/dplyr/vignettes/introduction.html).

I also prefer that stringsAsFactors = F be specified as an option at the start of any script, and generally avoid factors in these tables.

```{r, eval=FALSE}
library(scrattch)
library(dplyr)

options(stringsAsFactors=F)
```

```{r, include=FALSE}
library(scrattch)
library(dplyr)

options(stringsAsFactors=F)
```

-----

### <a name="data_section"></a> The *data* table

The data table contains the RPKM or count data for each gene from each sample. Because feather is very fast at retrieving columns, and we usually want to display data from only a few genes (but thousands of samples), this table should have genes as columns, and samples as rows.

The first column should be called "sample_id", with a unique name for each sample that matches the anno table, followed by columns named for each gene.

For convenience, the scrattch package now includes a function called flip_table(), which can be used to transpose data frame if samples are in columns instead of rows.

```{r}
rpkm <- v1_data[1:5,1:11]
rpkm

data <- flip_table(rpkm, gene_col = "gene", id_col = "sample_id")
data
```

[Back to Top](#top)

-----

### <a name="desc_section"></a> The *desc* Table

The *desc* table describes the annotations that are in the annotations (anno) table. This table contains the base of the annotation column names and the label that will be used for the annotation in the UI of our shiny pages. In addition, the desc table includes a type column, which ensures that shiny functions will display the annotations correctly. 

Currently, available types are: "cat" (for categorical data) and "num" (for numeric data).

For example, if the annotations are WGCNA clusters, Cre lines, and Number of Mapped Reads, this table could be easily made in R:

```{r}
desc <- data.frame(base=c("wgcna","cre","mapped"),
                   name=c("WGCNA Clusters","Cre Line","Number of Mapped Reads"),
                   type=c("cat", "cat", "num"))

desc
```

[Back to Top](#top)

-----

### <a name="anno_section"></a> The *anno* Table

The *anno* table contains the annotations that are used to group and arrange data from each cell in our Shiny apps.

In the *anno* table, each row represents a cell. The first column of the table contains cell IDs that match the column names of the *data* table.

Subsequent columns represent each annotation with three columns:

* _id: A numeric column that determines the order of the groups in the annotation
* _label: A character column that has the name of the annotation group, or the value of the annotation
* _color: A character column with the color of the annotation group as either a [hex color](http://www.color-hex.com/) or an [R color](http://www.stat.columbia.edu/~tzheng/files/Rcolor.pdf)

This table can be built in Excel as a csv, then imported into the database, or you can build it in R using the scrattch::annotate_cat() and scrattch::annotate_num() functions. This table is a representative set for the 10 cells included in the *data* table demo. If your table already has this format, it can be written directly to the feather directory:

```{r}
complete_anno <- v1_anno[1:10,]
head(complete_anno)
```

This table can also be built in R, or tables with partial data can be supplemented in R to include all three required columns. Below, a table with WGCNA clusters, Cre lines, and Mapped read values are expanded to include \_id and \_color columns.

```{r}
partial_anno <- demo_anno

head(partial_anno)
```

<a name="categorical"></a>To fill out the \_id and \_color columns, a summary table can be made for each annotation. These can be arranged as desired, colors can be added, and then the data can be joined together to make a complete *anno* table.

Below, I use the WGCNA Cluster labels to generate wgcna_id and wgcna_color data using scrattch's annotate_cat() function:
```{r}
anno <- partial_anno
head(anno)

anno <- anno %>%
  annotate_cat(col = "WGCNA.Cluster",
               base = "cluster",
               colorset = "rainbow")

head(anno)

```

<a name="quantitative"></a>For quantitative data, ids should be assigned based on the order in which you would like the data to be arranged, and colors can be assigned by choosing a position within a color ramp. This can all be handled automatically using scrattch's annotate_num() function:
```{r}
anno <- anno %>% 
  annotate_num(col = "Mapped.Reads",
               base = "mapped",
               scale = "zscore",
               colorset = c("dodgerblue","white","orangered"))

head(anno)
```

<a name ="manual"></a>Colors and ids could also be specified explicitly to define the order and color:
```{r}
cre_anno <- data.frame(cre_id = 1:3,
                       cre_label = c("Gad2", "Rorb", "Ntsr1"),
                       cre_color = c("orangered", "dodgerblue", "mediumorchid4"))

cre_anno
```


Now, all of these values can be joined together with the original, partial annotations to build the final anno table. Note that the mapped_label column has to now be converted to character from numeric so that it's used correctly by the functions:
```{r}
anno <- anno %>%
  rename(cre_label = Cre.Line) %>%
  left_join(cre_anno) %>%
  mutate(mapped_label = as.character(mapped_label))

head(anno)
```

This table is now complete and can be added to the feather directory It's not required, but you may want to arrange the table to group the annotation columns together:
```{r}
anno <- anno %>%
  group_annotations()

head(anno)
```

We now have all 3 required tables. These tables can be used interactively in scrattch, or can be written to disk for use in a Shiny app.

```{r}
dataset <- list(anno = anno,
                desc = desc,
                data = data)

group_violin_plot(data_source = dataset,
                  genes = names(data)[-1],
                  group_by = "cluster",
                  clusters = 1:6)

```

[Back to Top](#top)


-----

### <a name="db_write_section"></a> Writing the Database

Once you have all 3 required tables, you can use the convenient write_database() function provided by scrattch to build a database. We use a SQLite3 database because this makes accessing data from large datasets much faster than reading in an entire data set each time.

```{r, eval=FALSE}
write_database(desc = desc, anno = anno, data = rpkm, file = "demo.db")
```

This may take a few minutes depending in how much data is in the data table, as the table is indexed after writing for fast retrieval of rows using the "gene" column.

[Back to Top](#top)

-----

### <a name="db_udpate_section"></a> Updating a Database

Sometimes, you'll want to make a small change to the annotations or description tables without taking the time necessary to re-write and index the large data table. This can be accomplished using write_database as well by excluding arguments for the tables you don't want to update, and including the overwrite = T parameter:

```{r, eval=FALSE}
write_database(anno = new_anno, file = "demo.db", overwrite=T)
```

[Back to Top](#top)

-----

### <a name="db_check_section"></a> Check Database Structure

Once you've built a database, you may want to check its structure. scrattch includes a function to do just that: check_db_structure().

This function will make sure all of the cells in the data table have corresponding annotations in the anno table, that all of the annotations have the 3 required columns (id, label, and color), and that each item in the anno table is described in the desc table.

```{r, eval=FALSE}
check_db_structure("demo.db")
```

[Back to Top](#top)

